### PR TITLE
Working on keys-or-members unnesting

### PR DESCRIPTION
**Pull request by pavlopoul**

### CODE REVIEW

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/KeysOrMembersScalarEvaluator.java [hunk 1]
```diff
@@ -61,8 +64,11 @@ protected void evaluate(TaggedValuePointable[] args, IPointable result) throws S
                 case ValueTag.ARRAY_TAG:
                     abvs.reset();
                     sb.reset(abvs);
-                    tvp.getValue(ap);
-                    ap.appendItems(sb);
+                    while (keysOrMembers.step(tempTvp)) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/KeysOrMembersScalarEvaluator.java]: you need to call init before using the step function.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/KeysOrMembersUnnesting.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        }
+    }
+
+    public boolean step(IPointable result) throws SystemException {
+        if (arrayArgsLength > 0) {
+            while (indexArrayArgs < arrayArgsLength) {
+                ap.getEntry(indexArrayArgs, tvp);
+                abvs.reset();
+                try {
+                    ab.reset(abvs);
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/KeysOrMembersUnnesting.java]: I don't think you need to wrap the result in an array. Just return the individual value.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/KeysOrMembersScalarEvaluator.java [hunk 3]
```diff
@@ -30,6 +30,7 @@
 import org.apache.vxquery.exceptions.ErrorCode;
 import org.apache.vxquery.exceptions.SystemException;
 import org.apache.vxquery.runtime.functions.base.AbstractTaggedValueArgumentScalarEvaluator;
+import org.apache.vxquery.runtime.functions.step.ChildPathStepUnnesting;
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/KeysOrMembersScalarEvaluator.java]: I think this is the wrong import.

**Comment by prestoncarman** [comment]
I added a few comments, but this looks like your going in the right direction.

File: vxquery-core/src/main/java/org/apache/vxquery/jsonparser/JSONParser.java [hunk 4]
```diff
@@ -229,8 +233,13 @@ public int parse(Reader input, ArrayBackedValueStorage result) throws HyracksDat
 						obStack.get(levelObject - 1).finish();
 						if (itemStack.size() > 1) {
 							if (checkItem == itemType.OBJECT) {
-								obStack.get(levelObject - 2).addItem(spStack.get(levelObject - 2),
-										abvsStack.get(levelArray + levelObject));
+								if(levelObject > this.exactMatchLevel){
+									obStack.get(levelObject - 2).addItem(spStack.get(levelObject - 2),
+											abvsStack.get(levelArray + levelObject));
+								}else if(this.matched){
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/jsonparser/JSONParser.java]: Formatting.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java [hunk 5]
```diff
@@ -140,8 +141,16 @@
         // TODO Replace consolidate with a new child function that takes multiple paths.
         //        normalization.add(new ConsolidateUnnestsRule());
         normalization.add(new RemoveUnusedUnnestIterateRule());
+       // normalization.add(new RemoveRedundantVariablesRule());
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java:144]: Clean up comments.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java [hunk 6]
```diff
@@ -120,7 +122,9 @@ protected boolean processOperator(Mutable<ILogicalOperator> opRef, IOptimization
      */
     private boolean updateDataSource(VXQueryCollectionDataSource ds, Mutable<ILogicalExpression> expression) {
         boolean added = false;
+        //  FunctionIdentifier fi = null;
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushChildIntoDataScanRule.java:125]: Clean up comments.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushValueIntoDatascanRule.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+ *   ASSIGN( $v2 : constant )
+ *   DATASCAN( $source : $v1 )
+ *   plan__child
+ *
+ *   $source is encoded with the value parameters.
+ * </pre>
+ */
+
+public class PushValueIntoDatascanRule extends AbstractUsedVariablesProcessingRule {
+	StaticContext dCtx = null;
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushValueIntoDatascanRule.java]: Clean up indentation.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushValueIntoDatascanRule.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+		}
+		return false;
+	}
+
+	private boolean updateDataSource(VXQueryCollectionDataSource ds, Mutable<ILogicalExpression> expression) {
+		boolean added = false;
+		ILogicalExpression comparison = null;
+		List<Mutable<ILogicalExpression>> finds = new ArrayList<Mutable<ILogicalExpression>>();
+		ExpressionToolbox.findAllFunctionExpressions(expression, BuiltinOperators.VALUE.getFunctionIdentifier(), finds);
+		if (finds.size() > 0) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushValueIntoDatascanRule.java]: Can you explain what your doing here? What do the following variable do: comparison, b, and one?

**Review by prestoncarman — CHANGES_REQUESTED** [review]
Please fix the formatting on all the files (this will make it easier to see your changes). Also it would be helpful to remove any commented out code.

**Comment by prestoncarman** [comment]
What is the status of this change?

**Comment by pavlopoul** [comment]
This is an outdated pull request. The updated one is here: https://github.com/apache/vxquery/pull/163

**Pull request closed without merging**